onnx/backend/test/case/node/__init__.py

```diff
@@ -21,15 +22,60 @@
 _NodeTestCases = []
 _TargetOpType = ""
 
-from onnx.onnx_pb import NodeProto, AttributeProto, TypeProto, FunctionProto
+from onnx.onnx_pb import NodeProto, AttributeProto, TypeProto, FunctionProto, GraphProto
+
+
+def _rename_edges_helper(internal_node : NodeProto,
+                         rename_helper : Callable[[Text], Text],
+                         attribute_map : Dict[Text, Text]) -> NodeProto:
+    new_node = NodeProto()
+    new_node.CopyFrom(internal_node)
+    new_node.ClearField("input")
+    new_node.ClearField("output")
+    new_node.ClearField("attribute")
+    for internal_name in internal_node.input:
+        new_node.input.append(rename_helper(internal_name))
+    for internal_name in internal_node.output:
+        new_node.output.append(rename_helper(internal_name))
+    for attr in internal_node.attribute:
+        if attr.HasField("ref_attr_name"):
+            if attr.ref_attr_name in attribute_map:
+                new_attr = AttributeProto()
+                new_attr.CopyFrom(attribute_map[attr.ref_attr_name])  # type: ignore
+                new_attr.name = attr.name
+                new_node.attribute.extend([new_attr])
+        else:
+            new_attr = AttributeProto()
+            new_attr.CopyFrom(attr)
+            if attr.type == AttributeProto.GRAPHS:
+                new_graph = GraphProto()
+                new_graph.CopyFrom(attr.graphs[0])
+                for in_desc in new_graph.input:
+                    in_desc.name = rename_helper(in_desc.name)
```

Comment:
I think there is a tricky issue here. The current checker seems to allow a nested subgraph to have an input name same as a name from outer scope. (See here: https://github.com/onnx/onnx/blob/fb586d09aed80fa8529a87fdf01a6a98cd3619d5/onnx/checker.cc#L599 ). The comment there seems to question whether this should be allowed or not. But since it is currently allowed, the renaming needs to handle this case correctly.

Reply:
There is an inconsistency between the comments and the code. The code seems to allow subgraph inputs to shadow names from outer scope (the lines 597, 598, 599 etc.). While it is a good point that this should probably be cleaned up, changing it could potentially break existing models. Meanwhile, it seems better to use the code as the official spec, in case of conflicts like this.

Reply:
I think the issue with shadowed names should be addressed now